File: bindings/ruby/lib/cromulent.rb
# frozen_string_literal: true

# The Cromulent PRNG -- a pure-Ruby port of the scalar reference generator.
#
# {Cromulent::Engine} reproduces the identical 64-bit stream as the C reference
# implementation (+cromulent_init+ / +cromulent_next+) for any given seed;
# {Cromulent::StrongEngine} mirrors the heavier +cromulent_strong+ variant.
#
# Ruby integers are arbitrary precision, so every operation is masked back to
# 64 bits with MASK to reproduce the wrapping arithmetic of the C generator.
module Cromulent
  MASK = (1 << 64) - 1

  C1 = 0x9e3779b97f4a7c15
  C2 = 0xbf58476d1ce4e5b9
  C3 = 0x94d049bb133111eb
  C6 = 0xd1342543de82ef95
  MH3 = 0xd6e8feb86659fd93

  # Default seed, shared with the C library.
  DEFAULT_SEED = 0x853c49e6748fea9b

  module_function

  def rotl(x, k)
    ((x << k) | (x >> (64 - k))) & MASK
  end

  def mix_fast(x)
    x ^= x >> 32
    x = (x * MH3) & MASK
    x ^= x >> 32
    x
  end

  # SplitMix64-style seed expansion, matching cromulent_init.
  def seed_expand(seed)
    z = seed & MASK
    out = []
    2.times do
      z = (z + C1) & MASK
      z = ((z ^ (z >> 30)) * C2) & MASK
      z = ((z ^ (z >> 27)) * C3) & MASK
      out << (z ^ (z >> 31))
    end
    out
  end

  # The primary Cromulent generator.
  class Engine
    def initialize(seed = DEFAULT_SEED)
      @s0, @s1 = Cromulent.seed_expand(seed)
    end

    attr_reader :s0, :s1

    # Advance the state and return the next 64-bit output.
    def next_u64
      s0 = @s0
      s1 = @s1

      @s0 = (s0 * C6 + s1) & MASK
      @s1 = (Cromulent.rotl(s1, 31) + Cromulent.mix_fast(s0)) & MASK

      result = (s0 + Cromulent.rotl(s1, 11)) & MASK
      result ^= result >> 27
      result = (result * C3) & MASK
      result ^= result >> 27
      result
    end

    # Uniform float in [0, 1) using the top 53 bits.
    def next_float
      (next_u64 >> 11) * (1.0 / (1 << 53))
    end

    # Unbiased uniform integer in [0, n) via Lemire's method. Returns 0 when n == 0.
    def bounded(n)
      return 0 if n.zero?

      m = next_u64 * n
      low = m & MASK
      if low < n
        threshold = ((1 << 64) - n) % n
        while low < threshold
          m = next_u64 * n
          low = m & MASK
        end
      end
      m >> 64
    end

    # Advance the stream by z steps, discarding the output.
    def discard(z)
      z.times { next_u64 }
    end

    def ==(other)
      other.is_a?(Engine) && @s0 == other.s0 && @s1 == other.s1
    end
  end

  # The heavier "strong" Cromulent variant.
  class StrongEngine
    def initialize(seed = DEFAULT_SEED)
      @a, @b = Cromulent.seed_expand(seed)
    end

    attr_reader :a, :b

    def next_u64
      a = @a
      b = @b

      b = (b + Cromulent.rotl(a, 13)) & MASK
      a = (Cromulent.rotl(a, 29) * C1 + b) & MASK

      b = Cromulent.rotl(b, 17) ^ a
      a = (a + Cromulent.rotl((b * C2) & MASK, 31)) & MASK

      b = (b + Cromulent.rotl(a, 23)) & MASK
      a = Cromulent.rotl(a ^ b, 52)

      output = (a + Cromulent.rotl(b, 41)) & MASK

      @a = (a + C1) & MASK
      @b = b ^ (a >> 17)

      Cromulent.mix_fast(output)
    end

    def discard(z)
      z.times { next_u64 }
    end

    def ==(other)
      other.is_a?(StrongEngine) && @a == other.a && @b == other.b
    end
  end
end

File: bindings/python/cromulent.py
"""The Cromulent PRNG — a pure-Python port of the scalar reference generator.

``Engine`` produces the identical 64-bit stream as the C reference
implementation (``cromulent_init`` / ``cromulent_next``) for any given seed.
``StrongEngine`` mirrors the heavier ``cromulent_strong`` variant.

Python integers are arbitrary precision, so every operation is masked back to
64 bits with ``_MASK`` to reproduce the wrapping arithmetic of the C generator.
"""

from __future__ import annotations

__all__ = ["Engine", "StrongEngine", "DEFAULT_SEED"]

_MASK = (1 << 64) - 1

_C1 = 0x9E3779B97F4A7C15
_C2 = 0xBF58476D1CE4E5B9
_C3 = 0x94D049BB133111EB
_C6 = 0xD1342543DE82EF95
_MH3 = 0xD6E8FEB86659FD93

#: Default seed, shared with the C library.
DEFAULT_SEED = 0x853C49E6748FEA9B


def _rotl(x: int, k: int) -> int:
    return ((x << k) | (x >> (64 - k))) & _MASK


def _mix_fast(x: int) -> int:
    x ^= x >> 32
    x = (x * _MH3) & _MASK
    x ^= x >> 32
    return x


def _seed_expand(seed: int) -> tuple[int, int]:
    """SplitMix64-style seed expansion, matching cromulent_init."""
    z = seed & _MASK
    out = []
    for _ in range(2):
        z = (z + _C1) & _MASK
        z = ((z ^ (z >> 30)) * _C2) & _MASK
        z = ((z ^ (z >> 27)) * _C3) & _MASK
        out.append(z ^ (z >> 31))
    return out[0], out[1]


class Engine:
    """The primary Cromulent generator."""

    __slots__ = ("_s0", "_s1")

    def __init__(self, seed: int = DEFAULT_SEED) -> None:
        self._s0, self._s1 = _seed_expand(seed)

    def next_u64(self) -> int:
        """Advance the state and return the next 64-bit output."""
        s0 = self._s0
        s1 = self._s1

        self._s0 = (s0 * _C6 + s1) & _MASK
        self._s1 = (_rotl(s1, 31) + _mix_fast(s0)) & _MASK

        result = (s0 + _rotl(s1, 11)) & _MASK
        result ^= result >> 27
        result = (result * _C3) & _MASK
        result ^= result >> 27
        return result

    def random(self) -> float:
        """Uniform float in [0, 1) using the top 53 bits."""
        return (self.next_u64() >> 11) * (1.0 / (1 << 53))

    def next_float(self) -> float:
        """Uniform float in [0, 1) using the top 24 bits (single precision)."""
        return (self.next_u64() >> 40) * (1.0 / (1 << 24))

    def bounded(self, n: int) -> int:
        """Unbiased uniform integer in [0, n) via Lemire's method.

        Returns 0 when ``n == 0``.
        """
        if n == 0:
            return 0
        m = self.next_u64() * n
        low = m & _MASK
        if low < n:
            threshold = (-n) % n
            while low < threshold:
                m = self.next_u64() * n
                low = m & _MASK
        return m >> 64

    def discard(self, z: int) -> None:
        """Advance the stream by ``z`` steps, discarding the output."""
        for _ in range(z):
            self.next_u64()

    def __iter__(self):
        return self

    def __next__(self) -> int:
        return self.next_u64()

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Engine):
            return NotImplemented
        return self._s0 == other._s0 and self._s1 == other._s1


class StrongEngine:
    """The heavier "strong" Cromulent variant."""

    __slots__ = ("_a", "_b")

    def __init__(self, seed: int = DEFAULT_SEED) -> None:
        self._a, self._b = _seed_expand(seed)

    def next_u64(self) -> int:
        """Advance the state and return the next 64-bit output."""
        a = self._a
        b = self._b

        b = (b + _rotl(a, 13)) & _MASK
        a = (_rotl(a, 29) * _C1 + b) & _MASK

        b = _rotl(b, 17) ^ a
        a = (a + _rotl((b * _C2) & _MASK, 31)) & _MASK

        b = (b + _rotl(a, 23)) & _MASK
        a = _rotl(a ^ b, 52)

        output = (a + _rotl(b, 41)) & _MASK

        self._a = (a + _C1) & _MASK
        self._b = b ^ (a >> 17)

        return _mix_fast(output)

    def discard(self, z: int) -> None:
        for _ in range(z):
            self.next_u64()

    def __iter__(self):
        return self

    def __next__(self) -> int:
        return self.next_u64()

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, StrongEngine):
            return NotImplemented
        return self._a == other._a and self._b == other._b
